### ISSUE TITLE
Release binaries

### ISSUE THREAD

**Issue by bprashanth** [appc/cni#85]
It would make consumption a little easier, and why not anyway?

**Comment by tomdee** [comment]
I would also like binaries. It would be nice to have binaries attached to Github releases but it would also be great to have binaries per build.

Could it be done using the existing travis CI setup? I know that CircleCI allows the archiving of artifacts (we do this for Calico).

**Comment by bprashanth** [comment]
I think the stock CNI plugins *at least* need a Makefile that compile static binaries. 

IIUC there was some confusion around how exactly we should be distributing the plugins (https://groups.google.com/forum/#!topic/cni-dev/-fARxzl547s) but binaries are an easy enough first step that maybe we can start with that?

**Comment by tomdee** [comment]
Yes, agree strongly on the Makefile and static binaries, then that makes is easy to add them to CI builds or releases.

**Comment by errordeveloper** [comment]
One big +1 on building and publishing binaries in CI, although I'm not sure where we stand with tests...

**Comment by zachgersh** [comment]
@errordeveloper we are working on adding more test coverage to the whole project. Welcome PRs to increase it as well!

**Comment by tomdee** [comment]
[Travis](https://docs.travis-ci.com/user/uploading-artifacts/) supports uploading artifacts to S3 (CircleCI also supports hosting the artifacts themselves if you'd consider switching) -

**Comment by zachgersh** [comment]
Would be nicer if we could keep any sorts of binary artifacts as GitHub releases. This is usually easier in terms of discovery.

**Comment by tomdee** [comment]
Yes, binaries on Github releases would be a great thing, but 
1) this project doesn't have a great track record with doing releases on Github 
2) The github release binaries have to come from somewhere, and that somewhere should be reliable and repeatable. Having the CI system create the binaries would meet this requirement.

**Comment by philips** [comment]
Yea, we need to write some sort of release tooling. We have tooling for our other projects like rkt and etcd that we could borrow. If anyone has bandwidth to write the tooling we can do the releases and sign them with the CoreOS application signing key which has a whole trusted process around.

**Comment by bprashanth** [comment]
@kubernetes/sig-network

**Comment by zachgersh** [comment]
We may attempt to user [concourse](http://concourse.ci/) to test then build / upload our release binaries. Should be trivial to setup a pipeline (I've got plenty of experience). Would just need to find infrastructure to put it on (I am working on this).

cc @philips

**Comment by steveeJ** [comment]
@zachgersh Thanks for putting work into this! Any updates on this so far?

**Comment by zachgersh** [comment]
@steveeJ - yes, I've totally got a plan for this and was also waiting for [concourse](https://concourse.ci) to hit v1.0, which just recently happened.

Here are a couple of things I would love answers to:
- Can we get another repo in the appc org to host our pipeline / ci configuration?
- Where are we going to push the binaries to s3 wise for signing?
- Can someone provide the s3 credentials?
- Concourse runs on amazon. Would the team from CoreOS want to host this on their own Amazon account?

I am going to see what I can figure out on my side infrastructure wise - I may be able to host it on a shared account but that doesn't seem like a long-term solution.

Once we have those questions answered - we can totally get to work on setting up a pipeline.

**Comment by jonboulle** [comment]
Why does config need to be in a separate repo? 

I'll see about the aws stuff. Usually we have a preference for gcp. Does concourse run there?

**Comment by steveeJ** [comment]
> Why does config need to be in a separate repo?

Looking at the [concourse S3 resource type](https://github.com/concourse/s3-resource), I assume that it is to store the credentials somewhere safe, is this correct @zachgersh? Putting the credentials in one of the components system would be an alternative, but I couldn't find any support for this.

**Comment by tomdee** [comment]
What's the reason for using Concourse when travis is already being used? Maybe this is a naive question but I think it would be good to be explicit about it.

**Comment by zachgersh** [comment]
@tomdee great question - I find that concourse is much easier to debug failures in code versus having to figure out how to reproduce the travis environment you are running in (concourse I just jump into a docker image).

We could totally try this via travis upload directly to github releases. I wouldn't be opposed at all if it gets us to what we want :dancer: 

@steveeJ correct, we store that plus all of the other configuration of the tasks + pipelines in another repo so it doesn't get mixed in with the code that the pipeline is delivering (essentially a separation of concerns). cc @jonboulle

**Comment by steveeJ** [comment]
> correct, we store that plus all of the other configuration of the tasks + pipelines in another repo so it doesn't get mixed in with the code that the pipeline is delivering (essentially a separation of concerns)

Would it need to be a private repository? Are there alternatives?

**Comment by zachgersh** [comment]
The pipeline setup (YAML) could be public but the credentials have to go somewhere private.

**Comment by tomdee** [comment]
@zachgersh What are you anticipating will be in the Concourse pipeline? I see at least three statges to do a release 
1. Building static binaries
2. Signing those binaries
3. Publishing them as a Github release.

Given how infrequent releases are, and how sensitive the signing keys are, I'm not sure I see the value in automating steps 2 and 3 as part of a CI pipeline. WDYT?

**Comment by zachgersh** [comment]
@tomdee - these are the correct stages and with concourse I can automate 2 out of the three. I had imagined we would upload the binaries to an s3 bucket that was controlled by Coreos and someone who had the signing keys could sign them. Once they were signed, they could be placed into a different bucket where concourse would automatically see they were uploaded and push a release to github with said binaries. This feels like the best approach and gets us an automated process that the team doesn't have to give a ton of thought to as we go forward (the pipeline would be very simple).

cc @steveeJ

**Comment by tomdee** [comment]
We could 1) as a really small quick change - it's easy to upload to S3 from Travis. Then we would have binaries for every build. @steveeJ could then invoke the normal CoreOS signing process and just manually add the binaries to the release on Github. Given the current release cadence, I think it's better to get something out sooner rather than later.

I'm happy to submit a PR for the travis change if someone can add the S3 credentials to the travis environment.

**Comment by zachgersh** [comment]
@tomdee we totally decided to move this to the 0.3.0 milestone instead for 0.2.0 @steveeJ will manually sign and upload to github.

**Comment by luxas** [comment]
@steveeJ Can you comment here which steps you are taking when you're releasing manually?
I'm gonna cross-compile (`arm`, `arm64` and `ppc64le`) CNI for Kubernetes, so it would be really great to know how to build them for amd64 first...

**Comment by jonboulle** [comment]
@luxas we've been doing releases with `scripts/release-with-rkt.sh`

**Comment by bboreham** [comment]
Where are we on this now?
